### ISSUE TITLE
composing patterns

### ISSUE THREAD

**Issue by gjohnson** [tj/node-actorify#1]
Would be interesting to take some concepts from axon (reconnect, load-balance, etc) but apply them in a little more composable manner so the lib doesn't grow as much as axon did.  Obviously a little tricker if you leverage state within the actors... might be a good idea, might be a bad idea.

Concept:

``` js

var net = require('net');
var actorify = require('actorify');
var reconnect = require('reconnect');
var distribute = require('stream-distribute-thing');

// server

net.createServer(onconnect).listen(3000);
net.createServer(onconnect).listen(3001);

function onconnect (sock){
  var actor = actorify(sock);
  actor.on('ping', function(){
    console.log('PING');
    actor.send('pong');
  });
}

// client

var services = distribute([
  reconnect(3000),
  reconnect(3001)
]);

var actor = actorify(services);

setInterval(function(){
  actor.send('ping');
  actor.once('pong', function(){
    console.log('PONG');
  });
}, 300);


```

**Comment by tj** [comment]
yeah that's the plan! I don't think axon was a brutal size or anything but I think the actor abstraction usually makes more sense than more rpc-ish stuff, that's my major complaint with zmq/axon, but I would also never want to do all the boilerplate reconnection/loadbalance/actor stuff in an app, so we'll have to build a higher level actor thing on that. AND then I have some even bigger plans that also need to be a new layer

**Comment by tj** [comment]
node-amp now gives you opaque blobs instead of the lame object stream thing so that should make it easier to reuse some stream stuff if there's anything decent out there

**Comment by juliangruber** [comment]
immediately come to mind: mux-demux and reconnect

**Comment by stephen-fanfair** [comment]
If there was a reliable way to distribute stream messages like that, it'd be great. But it'd need to be aware of framing boundaries or you could potentially get part of a message sent to one endpoint and another part sent to another.

If you moved the framing codec to a point after the distributor elects a destination, it could work. You could add something like `services.codec(AMP)` to encode the data of each write call and pipe it to the internal distribution manager that jumps between the reconnect streams.
